app/controllers/pqs_controller.rb

```diff
@@ -11,28 +13,33 @@ def show
 
   def update
     loading_relations do
-      archive_trim_link!(params[:commit])
-
-      if @pq.update(pq_params)
-        PQProgressChangerService.new.update_progress(@pq)
-
-        action_officer_id = params.fetch(:commission_form, {})[:action_officer_id]
-
-        if action_officer_id.present?
-          @pq.reassign(ActionOfficer.find(action_officer_id))
+      with_valid_dates do
+        archive_trim_link!(params[:commit])
+
+        if @pq.update(pq_params) 
+          PQProgressChangerService.new.update_progress(@pq)
+          reassign_ao_if_present(@pq)
+          flash[:success] = 'Successfully updated'
+        else
+          @pq.trim_link(true)
+          flash[:error] = 'Update failed'
         end
 
-        flash[:success] = 'Successfully updated'
-        redirect_to action: 'show', id: @pq.uin
-      else
-        @pq.trim_link(true)
-        render action: 'show'
+        render :show
       end
     end
   end
 
   private
 
+  def with_valid_dates
+    date_keys.each { |key| pq_params[key].present? && parse_date(pq_params[key]) } 
```

Comment:
Do we need to through this column introspection business here? Perhaps it would be better to avoid the each loop here, and just call `date_parse` with the required columns as arguments